rain.js

```diff
@@ -0,0 +1,182 @@
+
+        const {intersect_test, points, leeway} = collider;
+        // For each vertex in that b, shift to the coordinate frame of
+        // a_inv*b.  Check if in that coordinate frame it penetrates
+        // the unit sphere at the origin.  Leave some leeway.
+        return points.arrays.position.some(p =>
+            intersect_test(T.times(p.to4(1)).to3(), leeway));
+    }
+}
+
+
+export class Simulation extends Scene {
+    // **Simulation** manages the stepping of simulation time.  Subclass it when making
+    // a Scene that is a physics demo.  This technique is careful to totally decouple
+    // the simulation from the frame rate (see below).
+    constructor() {
+        super();
+        Object.assign(this, {time_accumulator: 0, time_scale: 1, rain_enabled: true, t: 0, dt: 1 / 20, bodies: [], steps_taken: 0});
+    }
+
+    get rainEnabled() {
+        return this.rain_enabled;
+    }
+    simulate(frame_time) {
+        frame_time = this.time_scale * frame_time;
+        this.time_accumulator += Math.min(frame_time, 0.1);
+        while (Math.abs(this.time_accumulator) >= this.dt) {
+            this.update_state(this.dt);
+            for (let b of this.bodies)
+                b.advance(this.dt);
+            this.t += Math.sign(frame_time) * this.dt;
+            this.time_accumulator -= Math.sign(frame_time) * this.dt;
+            this.steps_taken++;
+        }
+        let alpha = this.time_accumulator / this.dt;
+        for (let b of this.bodies) b.blend_state(alpha);
+    }
+
+    make_control_panel() {
+        this.key_triggered_button("Toggle Rain", ["t"], () => this.rain_enabled = !this.rain_enabled);
+        this.key_triggered_button("Speed up time", ["Shift", "T"], () => this.time_scale *= 2);
+        this.key_triggered_button("Slow down time", ["t"], () => this.time_scale /= 2);
+        this.new_line();
+        this.live_string(box => {
+            box.textContent = "Time scale: " + this.time_scale
+        });
+        this.new_line();
+        this.live_string(box => {
+            box.textContent = "Fixed simulation time step size: " + this.dt
+        });
+        this.new_line();
+        this.live_string(box => {
+            box.textContent = this.steps_taken + " timesteps were taken so far."
+        });
+    }
+
+    display(context, program_state) {
+        if (program_state.animate)
+            this.simulate(program_state.animation_delta_time);
+        for (let b of this.bodies)
+            b.shape.draw(context, program_state, b.drawn_location, b.material);
+    }
+
+    update_state(dt) {
+        throw "Override this"
+    }
+}
+
+export class Test_Data {
+    constructor() {
+        this.textures = {
+            rgb: new Texture("assets/rgb.jpg"),
+            earth: new Texture("assets/earth.gif"),
+            grid: new Texture("assets/grid.png"),
+            stars: new Texture("assets/stars.png"),
+            text: new Texture("assets/text.png"),
+        }
```

Comment:
The assets folder is missing, but it doesn't matter since textures is not being used anywhere in the rain simulation. This block can be removed.
```suggestion
      
```